src/js/mqtt.js

```diff
@@ -65,60 +69,87 @@ MqttClient.prototype.connect = function() {
-      this.emit('connect');
-    } else if (res.type === MQTT_PUBLISH) {
-      var msg = this._handle._deserialize(res.buffer);
-      var ret = Object.assign(msg, {
-        toString: function(enc) {
-          return msg.payload.toString(enc || 'utf8');
-        }
-      });
-      this.emit('message', msg.topic, ret);
-      if (msg.qos > 0) {
-        // send publish ack
-        var ack = this._handle._getAck(msg.id, msg.qos);
-        this._write(ack, function(err) {
-          console.log('sent ack buffer with', err);
-        });
-      }
-    }
-  });
-  this._socket.on('error', (err) => {
-    if (this._socket && this._socket.end) {
-      this._socket.end();
-    }
-    if (this._reconnecting) {
-      if (this._reconnectingTimer)
-        clearTimeout(this._reconnectingTimer);
-      this._reconnectingTimer = setTimeout(
-        this.reconnect.bind(this), this._options.reconnectPeriod);
-    } else {
-      this.emit('error', err);
-    }
-  });
-  this._socket.on('end', (err) => {
-    this.emit('offline');
-    this.reconnect();
-  });
+  this._socket.on('data', this._ondata.bind(this));
+  this._socket.once('error', this._ondisconnect.bind(this));
+  this._socket.once('end', this._ondisconnect.bind(this));
+  this._lastConnectTime = Date.now();
   return this;
 };
 
 /**
  * @method _onconnect
  */
 MqttClient.prototype._onconnect = function() {
-  var buf = this._handle._getConnect();
-  return this._write(buf);
+  this._isConnected = true;
+  var buf;
+  try {
+    buf = this._handle._getConnect();
+  } catch (err) {
+    this.disconnect(err);
+    return;
+  }
+  this._write(buf);
+};
+
+MqttClient.prototype._onerror = function(err) {
+  this._ondisconnect(err);
+};
+
+MqttClient.prototype._onend = function() {
+  this._ondisconnect();
+};
+
+MqttClient.prototype._ondisconnect = function(err) {
+  if (err) {
+    this.emit('error', err);
+  }
+  if (this._isConnected) {
+    this._isConnected = false;
+    this.emit('offline');
+  }
+  this.reconnect();
+};
+
+MqttClient.prototype._ondata = function(chunk) {
+  var res;
+  try {
+    res = this._handle._readPacket(chunk);
+  } catch (err) {
+    this.disconnect(err);
+    return;
+  }
+  this.emit('packetreceive');
+
+  if (res.type === MQTT_CONNACK) {
+    if (this._reconnecting) {
+      clearTimeout(this._reconnectingTimer);
+      this._reconnecting = false;
+      this.emit('reconnect');
+    } else {
+      this.emit('connect');
+    }
+    this._keepAlive();
+  } else if (res.type === MQTT_PUBLISH) {
+    var msg;
+    try {
+      msg = this._handle._deserialize(res.buffer);
+    } catch (err) {
+      this.disconnect(err);
+      return;
+    }
+    this.emit('message', msg.topic, msg.payload);
+    if (msg.qos > 0) {
+      // send publish ack
+      try {
+        var ack = this._handle._getAck(msg.id, msg.qos);
+        this._write(ack);
+      } catch (err) {
+        this.disconnect(err);
+      }
+    }
+  } else {
+    // TODO handle other message type
```

Comment:
Needs throw some error or log something.

Reply:
i would like to log something here, but this will write a lot of messages to stdout after the program run a long time, is it a good choice?

Reply:
Emit another event `unhandled` would good to you?

Reply:
seems like a good idea